[No QA] Validate secondary logins in E.cash

NikkiWines opened the pull request:
### Details
Related to Expensify/Expensify.cash#2169 - right now when sending a user a validation link for a secondary login from E.cash we route them to E.com. This changes that so that we instead route them to E.cash where the validation is handled.

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes N/A

### Tests
Check out `nikki-secondarylogin-ecash-validationlink` in Web-Expensify
Follow test steps in https://github.com/Expensify/Web-Expensify/pull/30615

### QA Steps
<!--- 
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->
No QA, handled in https://github.com/Expensify/Web-Expensify/pull/30615

### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<details closed>
<summary>Web Videos</summary>
<br>

**Unauthenticated:**
- Valid Code 

   https://user-images.githubusercontent.com/16747822/114116023-c6a3b300-9898-11eb-9c8e-5c987994df54.mov

- Invalid Code (re-used an old validation link)

    https://user-images.githubusercontent.com/16747822/114080564-63962a00-9860-11eb-9de2-7e9aa60aa34c.mov

**Authenticated:**
- Valid Code 

    https://user-images.githubusercontent.com/16747822/114080521-57aa6800-9860-11eb-89bf-f3bc0d46a94c.mov

- Invalid Code (re-used an old validation link)

    https://user-images.githubusercontent.com/16747822/114080503-5416e100-9860-11eb-80f4-8161d91cb3e9.mov

</details>

#### Mobile Web
<details closed>
<summary>mWeb screenshots</summary>
<br>

**Unauthenticated:**
- Valid Code: 

    https://user-images.githubusercontent.com/16747822/114120829-218dd800-98a2-11eb-81b1-c9436cbc3413.mov

- Invalid Code: 

    https://user-images.githubusercontent.com/16747822/114120939-5bf77500-98a2-11eb-9a0d-7b21da8f5d2b.mov

**Authenticated:**
- Valid Code: 

    https://user-images.githubusercontent.com/16747822/114120970-67e33700-98a2-11eb-9828-cc75fb1640d4.mov

- Invalid Code:

    https://user-images.githubusercontent.com/16747822/114120977-6a459100-98a2-11eb-8810-ffad2785b71b.mov

</details>

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<details closed>
<summary>iOS screenshots</summary>
<br>

**Unauthenticated:**
- Valid Code: 

    https://user-images.githubusercontent.com/16747822/114118546-d671c600-989d-11eb-9dad-c0692f5c029f.mov

- Invalid Code: 

    https://user-images.githubusercontent.com/16747822/114118558-da054d00-989d-11eb-8a1c-2383d4a71ae0.mov

**Authenticated:**
- Valid Code: 

    https://user-images.githubusercontent.com/16747822/114118607-ef7a7700-989d-11eb-9724-f3ebb40eab2d.mov

- Invalid Code:

    https://user-images.githubusercontent.com/16747822/114118613-f2756780-989d-11eb-9166-0efc8feb0498.mov

</details>

#### Android
<details closed>
<summary>Android Videos</summary>
<br>

**Unauthenticated:**
- Valid Code 

    https://user-images.githubusercontent.com/16747822/114442934-f77a3580-9b81-11eb-95bf-07f0bec90984.mov

- Invalid Code (re-used an old validation link)

    https://user-images.githubusercontent.com/16747822/114442887-e8938300-9b81-11eb-9b24-adbe9a190b70.mov

**Authenticated:**
- Valid Code 

    https://user-images.githubusercontent.com/16747822/114442852-d9acd080-9b81-11eb-8ada-37e6ddb88abc.mov

- Invalid Code (re-used an old validation link)

    https://user-images.githubusercontent.com/16747822/114442777-c863c400-9b81-11eb-822a-8b79c1937412.mov

</details>

NikkiWines committed:
validate login when using v/accountID/validateCode url in e.cash

NikkiWines committed:
update logins if we do have an a session authtoken

NikkiWines commented:
This is kind of a weird flow in that we want it to run regardless of whether or not the user is authenticated. 

For the unauthenticated route, it made sense to route them to the login screen and then generate a temporary login if the request succeeds. Otherwise, we can display an error message on the homepage (shown above). 

For authenticated logins, I was debating the best thing to do (so there's no code for the route at the moment) - perhaps we could just check in the current URL matches the `v/accountID/validateCode` format and run the `validateLogin` logic if so? 

On a completely separate route, we could have a separate view (`ValidateLoginPage`) that we navigate to when we have a URL matching that format, which we could always navigate to whether the user is authenticated on not.

NikkiWines committed:
Merge branch 'master' of github.com:Expensify/Expensify.cash into nikki-validate-logins

NikkiWines committed:
re-add erroneously removed jsdoc

NikkiWines committed:
validate login when routed to AuthScreens.js

NikkiWines commented:
cc: @marcaaron if you have time since this involves some navigation stuff and I could use a sanity check

NikkiWines committed:
line length

NikkiWines committed:
redirect to settings page if user is already authenticated when clicking on their validation link

NikkiWines commented:
Still need to test this on other platforms but I think it's ready for a review

marcaaron commented on src/libs/actions/User.js:
```diff
@@ -103,11 +112,46 @@ function setSecondaryLogin(login, password) {
+                createTemporaryLogin(authToken, email);
+            } else {
+                fetch();
+                Navigation.navigate(ROUTES.SETTINGS_PROFILE);
```
Just clarifying the intention here. We do want to navigate immediately to the settings after the login is validated?

NikkiWines replied:
We don't have to, I just figured it would make sense because it would show that the secondary login now exists in a validated state in that view.

marcaaron replied:
Ah Ok I think that works unless someone said otherwise. Was just curious since it seems like this would also happen if you were validating your primary login (and not secondary).

But I might be confused here and perhaps it's not possible for an e.cash user to ever get a "magic sign" in link.

marcaaron commented on src/libs/actions/User.js:
```diff
@@ -103,11 +112,46 @@ function setSecondaryLogin(login, password) {
+ *
+ * @param {Number} accountID
+ * @param {String} validateCode
+ * @returns {Promise}
```
Didn't see anything using this response so assuming we maybe don't need to return anything from here.

marcaaron commented on src/libs/actions/User.js:
```diff
@@ -103,11 +112,46 @@ function setSecondaryLogin(login, password) {
+        return response;
+    }).finally((response) => {
+        Onyx.merge(ONYXKEYS.ACCOUNT, {loading: false});
+        return response;
```
Probably do not need to return the response unless something else will use it.

marcaaron commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -68,6 +70,14 @@ class AuthScreens extends React.Component {
+            // Split the /v/accountID/validateCode URL string into distinct components
+            // "", "v", "accountID", "validateCode"
+            const urlArray = this.props.currentURL.split('/');
+            User.validateLogin(Number(urlArray[2]), urlArray[3]);
```
It looks like this check is here so that we can handle the case of a user that is already authenticated. But if a user is already authenticated then presumably they are not logged in with the login they are trying to validate. 

Will weird stuff happen if we just validate this login anyway? e.g. do we need to clear out any stored data? Maybe force a sign out and then validate the login? Or something else? I haven't tested this yet so not sure what will happen but definitely curious.

NikkiWines replied:
I'm not sure I understand the flow you're describing here. This flow is for validating secondary logins so it's not possible for the user to be logged in with the login they're validating at this point.

marcaaron replied:
The flow would be...

1. User A logs in with primary account
2. User A creates a separate account (Note: Not secondary login) and gets a validate link for that
3. Since User A is logged in with the first account they created we will validate the account from the link and then change their auth token.

Maybe not a common occurrence, but I think it's weird that we can even do it. My expectation would be:

User A gets cleanly logged out of the first account they created and gets logged into the second account.

But maybe I'm misunderstanding this and that's not actually possible.

NikkiWines replied:
For validating primary accounts we use the `/setpassword/` URL so I think we would be fine, I will try testing this flow and try some edgecases for secondary logins to see what happens

marcaaron commented on src/libs/API.js:
```diff
@@ -400,6 +401,17 @@ function GetAccountStatus(parameters) {
+    const commandName = 'GetIOUReport';
+    requireParameters(['debtorEmail'], parameters, commandName);
+    return Network.post(commandName, parameters);
+}
```
NAB, curious why this is showing up in this PR since it seems unrelated

NikkiWines replied:
Ah, I moved them so that theyre in alphabetical order- which seemed to be the convention until recently. Happy to undo this change

marcaaron replied:
Ah ok, no worries really. Just seemed out of place in this PR + not sure if we are doing alphabetical order (I promise to definitely mess this up in the near future if we are).

NikkiWines replied:
ah yeah, I mean maybe if we just keep them inline with where they are in `export {` would be fine, but really anything goes - I end up cmd + f'ing to find them anyway so it doesn't really matter.

tgolen replied:
Alphabetical order was solely for the purpose of alleviating some of my OCD 😁

marcaaron commented on src/pages/signin/SignInPage.js:
```diff
@@ -44,16 +46,32 @@ const propTypes = {
     componentDidMount() {
+        // If we've redirected to the signin page with a validation code, validate the login
+        if (_.isMatch(this.props.route, {name: 'ValidateLogin'})) {
+            const accountID = Number(lodashGet(this.props.route, 'params.accountID', 0));
```
NAB/fun tip, if we feel like letting `react-navigation` do this for us we can parse this param and cast it to a `Number` directly in the `linkingConfig.js` then we can always expect it to be a number.

https://reactnavigation.org/docs/configuring-links/#serializing-and-parsing-params

marcaaron reviewed (commented):
This approach looks great to me. Just have a few initial comments.

tgolen commented on src/ROUTES.js:
```diff
@@ -17,8 +17,10 @@ export default {
     SEARCH: 'search',
     SIGNIN: 'signin',
-    SET_PASSWORD_WITH_VALIDATE_CODE: 'setpassword/:validateCode',
+    SET_PASSWORD_WITH_CODE: 'setpassword/:validateCode',
```
Hm, this is going to conflict with https://github.com/Expensify/Expensify.cash/pull/1857. We need to be careful there

NikkiWines replied:
I don't feel super strongly about naming conventions here, `VALIDATE_LOGIN_WITH_VALIDATE_CODE` just seemed kind of redundant. Happy to use that though to avoid conflicts.

tgolen commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -44,11 +44,13 @@ const RootStack = createCustomModalStackNavigator();
 
 const propTypes = {
     network: PropTypes.shape({isOffline: PropTypes.bool}),
+    currentURL: PropTypes.string,
```
These are missing inline comments for each prop, please add them

tgolen commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -68,6 +70,14 @@ class AuthScreens extends React.Component {
+            // Split the /v/accountID/validateCode URL string into distinct components
+            // "", "v", "accountID", "validateCode"
+            const urlArray = this.props.currentURL.split('/');
+            User.validateLogin(Number(urlArray[2]), urlArray[3]);
```
It's very possible that the code `Number(urlArray[2]), urlArray[3]` could break things if the URL is not exactly what it's expected to be. I'd suggest using something more like `lodash.get` to access the accountID and code.

NikkiWines replied:
ah yes, good point here.

tgolen commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -68,6 +70,14 @@ class AuthScreens extends React.Component {
+            // Split the /v/accountID/validateCode URL string into distinct components
+            // "", "v", "accountID", "validateCode"
+            const urlArray = this.props.currentURL.split('/');
+            User.validateLogin(Number(urlArray[2]), urlArray[3]);
```
I'm a bit concerned that `validateLogin()` is being called from `componentDidMount()`. That means that `AuthScreens` will render once, then validate the login, then redirect to the settings page if successful. I think what would be better to do is to prevent any rendering of `AuthScreens` at all until the validation API call is done. This could be done using a state variable and an Onyx key (maybe something in the session).

tgolen commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -68,6 +70,14 @@ class AuthScreens extends React.Component {
         }).then(subscribeToReportCommentEvents);
 
+        // If we have a validation code in our URL ensure we validate the associated login
+        if (this.props.currentURL.includes(`/${ROUTES.VALIDATE_LOGIN}/`)) {
```
Is there any reason why we couldn't use a specific page to handle the validation URLs? It seems like that is a better convention to follow (ie. each URL is handled by its own controller, allowing for a separation of concerns and keeping from having a single monolithic controller doing a bunch of different things.

NikkiWines replied:
I actually suggested something similar [here](https://github.com/Expensify/Expensify.cash/pull/2231#issuecomment-813602884) and it was what I first went to when building this out. However, we don't actually need any sort of view for this logic so I wasn't sure if it required its own page.

marcaaron replied:
We could maybe a component that just returns an `ActivityIndicator` and does the redirect logic when it mounts?

tgolen replied:
Yeah, I'd be fine with that. I think it makes all the route handling more intuitive and consistent, which is more important to me than having a component that renders very little (if anything at all).

tgolen commented on src/pages/signin/SignInPage.js:
```diff
@@ -44,16 +46,32 @@ const propTypes = {
 
 class SignInPage extends Component {
     componentDidMount() {
+        // If we've redirected to the signin page with a validation code, validate the login
```
Can the login be validated inside the action in the first place before redirecting? It seems strange to have an action redirect to a page, just to have the new page call another method in the same action.

NikkiWines replied:
Hmm so we navigate to SingInPage.js from [PublicScreens.js](src/libs/Navigation/AppNavigator/PublicScreens.js)... I suppose we don't need to redirect here to do that and instead could it in PublicScreens.js. 

To be honest, I'm not sure what the best flow is here - ideally, we'd run this immediately since we want it to happen regardless of whether or not the user is currently logged in, but we don't have any other logic like that at the moment (as far I know of). 

Definitely open to exploring other implementations

tgolen replied:
Ah, OK. Sorry, I misunderstood the redirect from `User.validateLogin()`.

I think that `PublicScreens.js` should have a route handling specifically the validate login functionality, then once that's done, let it redirect to SignIn. Does that make sense? I think that's inline with the other comments about using explicit routes.

tgolen commented on src/pages/signin/SignInPage.js:
```diff
@@ -44,16 +46,32 @@ const propTypes = {
+        if (_.isMatch(this.props.route, {name: 'ValidateLogin'})) {
+            const accountID = Number(lodashGet(this.props.route, 'params.accountID', 0));
+            const validateCode = lodashGet(this.props.route, 'params.validateCode', '');
+            validateLogin(accountID, validateCode);
```
I'm confused about why `validateLogin` is being called a second time. I thought it was already called in `AuthScreens`, then it validates the login, and redirects to the settings page. Can you help me understand the code flow and what I'm missing?

NikkiWines replied:
Ah, yes, so there's two routes for validation here depending on what state the user is in : 

**1. User is not signed in when clicking on validation link:** 
 - We navigate to PublicScreens.js where we see that we have a public route matching `ValidateLogin` and redirect to the signin page where we run `User.validateLogin`

**2. User is already logged in when clicking on the validation link**
- We go directly to AuthScreens.js, and run `User.validateLogin` in `componentDidMount()`, we redirect the user to Settings > Profile to show that their login has been updated

NikkiWines commented on src/libs/actions/User.js:
```diff
@@ -103,11 +112,46 @@ function setSecondaryLogin(login, password) {
+            if (!sessionAuthToken) {
+                createTemporaryLogin(authToken, email);
+            } else {
+                fetch();
```
Ah yeah, I can update this to `getUserDetails()` or something

NikkiWines committed:
Merge branch 'master' of github.com:Expensify/Expensify.cash into nikki-validate-logins

NikkiWines committed:
modify validateLogin logic to go through ValidateLoginPage.js

NikkiWines commented:
Ok made some changes that I _think_ are more in line with what we were discussing re: routing things through one component. 

The only thing is it doesn't seem to like redirecting to anything if the user is signed in. Redirect logic works fine if the user is signed out.

tgolen commented on src/libs/actions/User.js:
```diff
@@ -119,11 +128,47 @@ function setSecondaryLogin(login, password) {
+ * @param {String} validateCode
+ */
+function validateLogin(accountID, validateCode) {
+    let isLoggedIn = sessionAuthToken;
```
`isLoggedIn` implies a boolean, so I would suggest either casting to a boolean, or change the name.

tgolen commented on src/pages/ValidateLoginPage.js:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    render() {
+        return null;
```
Can you please add a comment here to explain why nothing is being rendered (just explain that this page is handling the validate route and will redirect to the proper page after it's done)?

tgolen commented on src/pages/routePropTypes.js:
```diff
@@ -0,0 +1,10 @@
+    route: PropTypes.shape({
+        // Each parameter passed via the URL
+        params: PropTypes.shape({
+        }),
```
Missing some information in this shape?

NikkiWines replied:
Ah I left this intentionally vague so it could be used by other components, but you're right that we should specify that the params are accountID and validateCode somewhere.

tgolen reviewed (changes requested):
I love the way that cleaned up the logic. It's MUCH easier to follow now, thank you!

What do we need to do about that redirect problem? Maybe @marcaaron has some ideas of why it's not working? Did you dig into it yet to see why it's not working?

NikkiWines commented:
> What do we need to do about that redirect problem? Maybe @marcaaron has some ideas of why it's not working? Did you dig into it yet to see why it's not working?

I dug into it a little yesterday without success and plan to keep looking at it today some more to see what's happening there!

NikkiWines committed:
redirec to currentlyViewedReportID instead of home so that we display the view correctly after validating

NikkiWines commented:
Amazing what looking at things with fresh eyes can do. Turns out the issue was that navigation _was_ working (the URL wasn't updating though which is weird) but we didn't have anything to display at `""`. Navigating to the currently viewed reportID fixed this issue! 

Here's an example with an invalid/already used code when the user is logged in.

https://user-images.githubusercontent.com/16747822/114076693-d0f38c00-985b-11eb-87ca-45de45e94e87.mov

NikkiWines committed:
fix import

NikkiWines committed:
update validationLinkPropTypes

NikkiWines commented:
Hmm ok, I was running into an issue where signing in a previously unauthenticated user after validating their login took us back to `ValidateLoginPage'. So, I tried setting it so we didn't try to log the user in after a successful validation... 

That worked... sort of. We do navigate back to the `signin` screen, but then when the user logs (like normal) in we somehow _still_ get looped back to `ValidateLoginPage`. It looks like somehow the `initialURL` retrieved by `Linking.getInitialURL()` is still `v/accountID/validateCode` instead of `signin` after redirecting. 

I'm at a bit of a loss on how to fix that... 😓 

Here's a video of that issue in action: 

https://user-images.githubusercontent.com/16747822/114101647-de6c3e80-987a-11eb-9550-43029b4ee229.mov

NikkiWines committed:
don't try to log the user in after validating secondary login

marcaaron commented on src/libs/actions/User.js:
```diff
@@ -119,11 +134,48 @@ function setSecondaryLogin(login, password) {
+        }
+    }).finally(() => {
+        Onyx.merge(ONYXKEYS.ACCOUNT, {loading: false});
+        setCurrentURL(redirectRoute);
```
`setCurrentURL()` happens automatically and is just a side effect of navigating

https://github.com/Expensify/Expensify.cash/blob/7300c61f7ccc4399dff7e8d866ae3b52cb6b965d/src/libs/Navigation/NavigationRoot.js#L98-L112

marcaaron commented on src/libs/actions/User.js:
```diff
@@ -119,11 +134,48 @@ function setSecondaryLogin(login, password) {
+        if (response.jsonCode === 200) {
+            const {email} = response;
+
+            // If the user is unauthenticated, generate a login for them with the returned authtoken and sign then in
```
Out of date comment I think since we are redirecting to sign in with a success/failure message

NikkiWines committed:
update comment, remove unecessary setCurrentURL

marcaaron commented:
Testing a bit on iOS and noticed that the issue you're running into @NikkiWines doesn't happen there. So that's something 😅 

https://user-images.githubusercontent.com/32969087/114108150-381a3c00-986e-11eb-88d0-da361860fd45.mp4

There's kind of a weird slide in and out animation effect happening though that I'm not sure we want.

Here's my current theory though not sure if it's correct.

- When we first navigate to `/v` we push a new screen onto the stack (as shown on iOS above)
- We redirect to `/signin` and now we are sitting on top of the `/v` page (which renders nothing)
- After we authenticate we are dropped back into the auth screens and since we just have one route and it references `ValidateLogin` (which is also duplicated in `AuthScreens`) that's what we'll see when we get logged in

That doesn't explain why it works on mobile and not web... but I do see an error on web. These logs pretty much verify most of the theory. Except it looks like we are trying to redirect to `/r` on web and fail to do so for some reason.

![Screen Shot 2021-04-08 at 1 32 01 PM](https://user-images.githubusercontent.com/32969087/114109227-8cbeb680-9870-11eb-8633-e965331b57a7.png)

marcaaron commented:
OH yeah I think that we redirect to `/r` because we try to validate the login again and then redirect to a `currentlyViewedReportID` that doesn't exist yet because the user is not properly logged in. Ideally, we would just not land back in `/v` again though I think.

marcaaron commented:
The default navigate behavior in the stack navigator is going to be a "push", but we can change that behavior by making it do a "replace". Or we can probably also make the main public routes stack use mode modal.

**Solution 1:**

In `Navigation.navigate()` do something like this 

```js
    if (route === ROUTES.SIGNIN) {
        navigationRef.current.dispatch(StackActions.replace('SignIn'));
        return;
    }
```

**Solution 2:**

This is kind of fiddly. But when navigating to the `/v` link and unauthenticated we could set up the navigation state so that the `SignIn` screen is underneath the `ValidateLogin` screen (basically setting a specific state in the stack). Then instead of redirecting to `SignIn` we can just pop off the `ValidateLogin` screen instead.

There might be another way, but that's what I got for now.

NikkiWines commented:
Oh super interesting. I like **Solution 1** a lot since it's pretty minimal and I don't think there's a flow where we'd want to retain things when navigating to`signIn`. I'll try implementing that now!

NikkiWines committed:
add success message to narrow signin view

remove @expensify.sms domain for ChangeExpensifyLoginLink

NikkiWines commented:
Updated! Everything works and i've added videos for Web + iOS. Working on android and mWeb now. Not sure how I'll test desktop but I'm sure we can figure something out.

tgolen commented on src/CONST.js:
```diff
@@ -77,6 +77,7 @@ const CONST = {
         TIMEZONE: 'timeZone',
     },
     DEFAULT_TIME_ZONE: {automatic: true, selected: 'America/Los_Angeles'},
+    DEFAULT_ACCOUNT_DATA: {error: '', success: '', loading: true},
```
Hm, this kind of gives me a thought, and I am fine with it being a NAB, but I think there are other places that are using this same pattern. I'm primarily thinking about these:

https://github.com/Expensify/Expensify.cash/blob/a740e888cc47b7980b37d7c7c8366d2096fca03f/src/Expensify.js#L22-L26

Which also makes me wonder if that some code should be referencing this constant too?

marcaaron replied:
I like this thought. Would also love to see a follow up issue or maybe discussion in `#expensify-open-source` to get more ideas on how to standardize API loading and error cases. Not convinced what we are doing now is always clear or easy to follow, but works for now.

NikkiWines replied:
Updated Expensify.js to use `DEFAULT_ACCOUNT_DATA` for `[ONYXKEYS.ACCOUNT]`

tgolen commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -97,7 +104,7 @@ class AuthScreens extends React.Component {
                 return;
```
I know that this isn't code from this PR, so I can consider it another NAB, but I think this is not quite done the proper way. The way I would expect this to happen is to use `componentDidUpdate` and if the network prop changed to being offline, then the interval would be cleared. Then when the network prop changes to online again, the interval would be re-created.

NikkiWines replied:
Hmm, not sure I fully understand your suggestion so perhaps it would be better to have this change in a separate PR

tgolen replied:
OK, that works. I'll maybe submit something to see if my idea works.

tgolen commented on src/libs/Navigation/Navigation.js:
```diff
@@ -33,6 +33,12 @@ function goBack() {
  * @param {String} route
  */
 function navigate(route = ROUTES.HOME) {
+    // If we're navigating to the signIn page, replace the existing route in the stack with the SignIn route.
```
Can you please modify this comment to explain "why" this is being done?

tgolen commented on src/pages/ValidateLoginPage.js:
```diff
@@ -0,0 +1,34 @@
+    },
+};
+
+export default class ValidateLoginPage extends Component {
```
The export statement needs to be moved to the end of the file.

tgolen commented:
Yeah, I was thinking the same... It sounds like these cases are happening
more and more so getting them standardized in some fashion feels right.

On Fri, Apr 9, 2021 at 12:00 PM Marc Glasser ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/CONST.js
> <https://github.com/Expensify/Expensify.cash/pull/2231#discussion_r610813868>
> :
>
> > @@ -77,6 +77,7 @@ const CONST = {
>          TIMEZONE: 'timeZone',
>      },
>      DEFAULT_TIME_ZONE: {automatic: true, selected: 'America/Los_Angeles'},
> +    DEFAULT_ACCOUNT_DATA: {error: '', success: '', loading: true},
>
> I like this thought. Would also love to see a follow up issue or maybe
> discussion in #expensify-open-source to get more ideas on how to
> standardize API loading and error cases. Not convinced what we are doing
> now is always clear or easy to follow, but works for now.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Expensify/Expensify.cash/pull/2231#discussion_r610813868>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJMAB5WVZYZYUGATPDZR5LTH46FDANCNFSM42NKLH6Q>
> .
>

NikkiWines committed:
update comments, use const where applicable, and put export at EOF

NikkiWines committed:
Merge branch 'master' of github.com:Expensify/Expensify.cash into nikki-validate-logins

NikkiWines committed:
update default account data const to not inlcude loading

NikkiWines commented:
Still having issues with the links for android, but the code should be ready for a final review

NikkiWines committed:
Merge branch 'master' of github.com:Expensify/Expensify.cash into nikki-validate-logins

NikkiWines commented:
Updated with Android videos

tgolen commented on src/CONST.js:
```diff
@@ -77,6 +77,7 @@ const CONST = {
         TIMEZONE: 'timeZone',
     },
     DEFAULT_TIME_ZONE: {automatic: true, selected: 'America/Los_Angeles'},
+    DEFAULT_ACCOUNT_DATA: {error: '', success: ''},
```
Since this is the default state, why not have `loading: false` in it?

tgolen commented on src/libs/actions/Session.js:
```diff
@@ -224,7 +225,7 @@ function restartSignin() {
  */
 function setPassword(password, validateCode, accountID) {
-    Onyx.merge(ONYXKEYS.ACCOUNT, {error: '', loading: true});
+    Onyx.merge(ONYXKEYS.ACCOUNT, {...CONST.DEFAULT_ACCOUNT_DATA, ...{loading: true}});
```
I think these can all be a little more simple:
```suggestion
    Onyx.merge(ONYXKEYS.ACCOUNT, {...CONST.DEFAULT_ACCOUNT_DATA, loading: true});
```

I tried it in the console and it looks like it works:
![image](https://user-images.githubusercontent.com/1228807/114465046-ce6a9c80-9ba3-11eb-997b-aacd9fe84794.png)

NikkiWines committed:
tweaks to default_account_data declaration and usage

NikkiWines commented:
updated!

NikkiWines committed:
dont use {...} where unnecessary

marcaaron commented on src/pages/signin/ChangeExpensifyLoginLink.js:
```diff
@@ -24,7 +25,8 @@ const ChangeExpensifyLoginLink = ({credentials}) => (
             <Text style={[styles.link]}>
                 Not&nbsp;
-                {credentials.login}
+                {Str.removeSMSDomain(credentials.login)}
```
Heads up there there is an open issue for this

https://github.com/Expensify/Expensify.cash/issues/2323

NikkiWines replied:
Ah, oops. Didn't think to look for open issues, just happened upon it in testing and figured it was a super easy fix.

NikkiWines replied:
Should I remove this line so that the contributor who has already put in the work to fix it can still get paid?

marcaaron commented on src/pages/signin/LoginForm/LoginFormWide.js:
```diff
@@ -90,6 +93,11 @@ class LoginFormWide extends React.Component {
+                    <Text style={[styles.formSuccess]}>
+                        {this.props.account.success}
+                    </Text>
+                )}
```
cc @jasperhuangg since we were discussing a refactor to combine the wide and narrow versions. See how we needed to update this in two places for this PR?

jasperhuangg replied:
Yup!! Thanks for the tag; great to know how the cleanup will help :)

marcaaron reviewed (approved):
LGTM
